// @vxe-ui/core/lib/src/mixins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sizeMixin = exports.permissionMixin = exports.globalMixins = void 0;
var _vue = _interopRequireDefault(require("vue"));
var _permission = require("./permission");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const sizeMixin = exports.sizeMixin = _vue.default.extend({
  inject: {
    $xeSizeInfo: {
      default: null
    }
  },
  provide() {
    return {
      $xeSizeInfo: {
        getSize: () => this.computeSize
      }
    };
  },
  computed: {
    computeSize() {
      const {
        size
      } = this;
      const $xeSizeInfo = this.$xeSizeInfo;
      return size || ($xeSizeInfo ? $xeSizeInfo.getSize() : null);
    }
  }
});
const permissionMixin = exports.permissionMixin = _vue.default.extend({
  computed: {
    computePermissionInfo() {
      return (0, _permission.handleCheckInfo)(this.permissionCode, this.permissionMethod);
    }
  }
});
const globalMixins = exports.globalMixins = {
  sizeMixin,
  permissionMixin
};
