// vxe-pc-ui/lib/loading/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.VxeLoading = exports.LoadingController = exports.Loading = void 0;
var _core = require("@vxe-ui/core");
var _loading = _interopRequireDefault(require("./src/loading"));
var _dynamics = require("../dynamics");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var VxeLoading = exports.VxeLoading = Object.assign({}, _loading.default, {
  install: function install(app) {
    app.component(_loading.default.name, _loading.default);
  }
});
var LoadingController = exports.LoadingController = {
  open: function open(options) {
    var opts = Object.assign({}, options);
    _dynamics.dynamicStore.globalLoading = {
      value: true,
      text: opts.text,
      icon: opts.icon
    };
    (0, _dynamics.checkDynamic)();
  },
  close: function close() {
    _dynamics.dynamicStore.globalLoading = null;
  }
};
_dynamics.dynamicApp.use(VxeLoading);
_core.VxeUI.component(_loading.default);
_core.VxeUI.loading = LoadingController;
var Loading = exports.Loading = VxeLoading;
var _default = exports.default = VxeLoading;
