// vxe-pc-ui/lib/radio/src/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _comp = require("../../ui/src/comp");
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _ui = require("../../ui");
var _radio = _interopRequireDefault(require("./radio"));
var _button = _interopRequireDefault(require("./button"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var _default2 = exports.default = /* define-vxe-component start */(0, _comp.defineVxeComponent)({
  name: 'VxeRadioGroup',
  mixins: [_ui.globalMixins.sizeMixin, _ui.globalMixins.permissionMixin],
  props: {
    value: [String, Number, Boolean],
    disabled: {
      type: Boolean,
      default: null
    },
    type: String,
    options: Array,
    optionProps: Object,
    strict: {
      type: Boolean,
      default: function _default() {
        return (0, _ui.getConfig)().radioGroup.strict;
      }
    },
    size: {
      type: String,
      default: function _default() {
        return (0, _ui.getConfig)().radioGroup.size || (0, _ui.getConfig)().size;
      }
    }
  },
  inject: {
    $xeForm: {
      default: null
    },
    formItemInfo: {
      from: 'xeFormItemInfo',
      default: null
    }
  },
  provide: function provide() {
    var $xeRadioGroup = this;
    return {
      $xeRadioGroup: $xeRadioGroup
    };
  },
  data: function data() {
    var reactData = {};
    return {
      xID: _xeUtils.default.uniqueId(),
      name: _xeUtils.default.uniqueId('xe_group_'),
      reactData: reactData
    };
  },
  computed: Object.assign(Object.assign({}, {}), {
    computeIsDisabled: function computeIsDisabled() {
      var $xeRadioGroup = this;
      var props = $xeRadioGroup;
      var $xeForm = $xeRadioGroup.$xeForm;
      var disabled = props.disabled;
      if (disabled === null) {
        if ($xeForm) {
          return $xeForm.readonly || $xeForm.disabled;
        }
        return false;
      }
      return disabled;
    },
    computePropsOpts: function computePropsOpts() {
      var $xeRadioGroup = this;
      var props = $xeRadioGroup;
      return Object.assign({}, props.optionProps);
    },
    computeLabelField: function computeLabelField() {
      var $xeRadioGroup = this;
      var propsOpts = $xeRadioGroup.computePropsOpts;
      return propsOpts.label || 'label';
    },
    computeValueField: function computeValueField() {
      var $xeRadioGroup = this;
      var propsOpts = $xeRadioGroup.computePropsOpts;
      return propsOpts.value || 'value';
    },
    computeDisabledField: function computeDisabledField() {
      var $xeRadioGroup = this;
      var propsOpts = $xeRadioGroup.computePropsOpts;
      return propsOpts.disabled || 'disabled';
    }
  }),
  methods: {
    //
    // Method
    //
    dispatchEvent: function dispatchEvent(type, params, evnt) {
      var $xeRadioGroup = this;
      $xeRadioGroup.$emit(type, (0, _ui.createEvent)(evnt, {
        $radioGroup: $xeRadioGroup
      }, params));
    },
    emitModel: function emitModel(value) {
      var $xeRadioGroup = this;
      var _events = $xeRadioGroup._events;
      $xeRadioGroup.$emit('input', value);
      if (_events && _events.modelValue) {
        $xeRadioGroup.$emit('modelValue', value);
      } else {
        $xeRadioGroup.$emit('model-value', value);
      }
    },
    handleChecked: function handleChecked(params, evnt) {
      var $xeRadioGroup = this;
      var $xeForm = $xeRadioGroup.$xeForm;
      var formItemInfo = $xeRadioGroup.formItemInfo;
      var value = params.label;
      $xeRadioGroup.emitModel(value);
      $xeRadioGroup.dispatchEvent('change', {
        value: value,
        label: value
      }, evnt);
      // 自动更新校验状态
      if ($xeForm && formItemInfo) {
        $xeForm.triggerItemEvent(evnt, formItemInfo.itemConfig.field, value);
      }
    },
    //
    // Render
    //
    renderVN: function renderVN(h) {
      var $xeRadioGroup = this;
      var props = $xeRadioGroup;
      var slots = $xeRadioGroup.$scopedSlots;
      var options = props.options,
        type = props.type;
      var vSize = $xeRadioGroup.computeSize;
      var defaultSlot = slots.default;
      var valueField = $xeRadioGroup.computeValueField;
      var labelField = $xeRadioGroup.computeLabelField;
      var disabledField = $xeRadioGroup.computeDisabledField;
      var btnComp = type === 'button' ? _button.default : _radio.default;
      return h('div', {
        class: ['vxe-radio-group', _defineProperty({}, "size--".concat(vSize), vSize)]
      }, defaultSlot ? defaultSlot({}) : options ? options.map(function (item) {
        return h(btnComp, {
          key: item[valueField],
          props: {
            label: item[valueField],
            content: item[labelField],
            disabled: item[disabledField]
          }
        });
      }) : []);
    }
  },
  render: function render(h) {
    return this.renderVN(h);
  }
});
/* define-vxe-component end */
