// vxe-pc-ui/lib/select/src/util.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assembleOption = assembleOption;
exports.createOption = createOption;
exports.destroyOption = destroyOption;
exports.isOption = isOption;
exports.watchOption = watchOption;
var _xeUtils = _interopRequireDefault(require("xe-utils"));
var _optionInfo = require("./option-info");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function isOption(option) {
  return option instanceof _optionInfo.OptionInfo;
}
function createOption($xeSelect, _vm) {
  return isOption(_vm) ? _vm : new _optionInfo.OptionInfo($xeSelect, _vm);
}
function watchOption(_vm, option) {
  var optionPropKeys = ['value', 'label', 'visible', 'className', 'disabled'];
  optionPropKeys.forEach(function (name) {
    _vm.$watch(name, function (val) {
      option.update(name, val);
    });
  });
}
function assembleOption($xeSelect, el, option, xeOptgroup) {
  var reactData = $xeSelect.reactData;
  var staticOptions = reactData.staticOptions;
  var parentElem = el.parentNode;
  var parentOption = xeOptgroup ? xeOptgroup.optionConfig : null;
  var parentCols = parentOption ? parentOption.options : staticOptions;
  if (parentElem && parentCols) {
    parentCols.splice(_xeUtils.default.arrayIndexOf(parentElem.children, el), 0, option);
    reactData.staticOptions = staticOptions.slice(0);
  }
}
function destroyOption($xeSelect, option) {
  var reactData = $xeSelect.reactData;
  var staticOptions = reactData.staticOptions;
  var matchObj = _xeUtils.default.findTree(staticOptions, function (item) {
    return item.id === option.id;
  }, {
    children: 'options'
  });
  if (matchObj) {
    matchObj.items.splice(matchObj.index, 1);
  }
  reactData.staticOptions = staticOptions.slice(0);
}
